File: Alignment/family_alignment.cpp
//Faraz Doagooye Tehrani

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <climits>

using namespace std;

int score(char first, char second) {
    return (first == second) ? 1 : -1;
}

vector< vector<int> > ga(const string &v, const string &w) {
    int rows = v.size() + 1;
    int cols = w.size() + 1;
    vector< vector<int> > S(rows, vector<int>(cols, INT_MIN/2));
    
    S[0][0] = 0;
    for (int i = 1; i < rows; i++) S[i][0] = -i;
    for (int j = 1; j < cols; j++) S[0][j] = -j;
    
    for (int i = 1; i < rows; i++) {
        for (int j = 1; j < cols; j++) {
            S[i][j] = max(max(
                S[i-1][j-1] + score(v[i-1], w[j-1]),
                S[i-1][j] - 1),
                S[i][j-1] - 1
            );
        }
    }
    return S;
}

int main() {
    string first;
    getline(cin, first);
    
    int n;
    cin >> n;
    cin.ignore();
    
    vector<string> strings(n);
    for (int i = 0; i < n; i++)
        getline(cin, strings[i]);
    
    int maxNum = INT_MIN;
    
    vector< vector< vector<int> > > prefixScores(n);
    vector< vector < vector<int> > > suffixScores(n);
    
    for (int i = 0; i < n; i++) {
        prefixScores[i] = ga(first, strings[i]);
    }
    
    string revFirst(first.rbegin(), first.rend());
    for (int i = 0; i < n; i++) {
        string revStr(strings[i].rbegin(), strings[i].rend());
        suffixScores[i] = ga(revFirst, revStr);
    }
    
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            int len1 = strings[i].size();
            int len2 = strings[j].size();
            int m = first.size();
            
            for (int k = 0; k <= m; k++) {
                int score1 = prefixScores[i][k][len1];
                int score2 = suffixScores[j][m-k][len2];
                maxNum = max(maxNum, score1 + score2);
            }
        }
    }
    
    cout << maxNum << "\n";
    return 0;
}

File: HW2/family_alignment.py
#Faraz Doagooye Tehrani 402105998

def score(first, second):
    if first == second:
        return 1
    return -1

def ga(v, w):
    rows = len(v) + 1
    cols = len(w) + 1

    S = [[0 for _ in range(cols)] for _ in range(rows)]  
    for i in range(len(S)):
        S[i][0] = -1 * i
    for i in range(len(S[0])):
        S[0][i] = -1 * i

    for i in range(1, len(S)):
        for j in range(1, len(S[0])):
            S[i][j] = max(S[i - 1][j] - 1, S[i][j - 1] - 1, 
                           S[i - 1][j - 1] + score(v[i - 1], w[j - 1]))

    return S

            
first = input().strip()
maxNum = [float('-inf')] * (len(first) + 1)
maxNum_Reversed = [float('-inf')] * (len(first) + 1)
maxFinal = float('-inf')
num = int(input())

strings = []
result = []
result_rev = []
for i in range(num):
    strings.append(input().strip())

for string in strings:
    result = ga(first, string)
    result_rev = ga(''.join(reversed(first)), ''.join(reversed(string)))



for i in range(1, len(first) + 1):
    if maxNum[i] + maxNum_Reversed[len(first) - i] > maxFinal:
        maxFinal = maxNum[i] + maxNum_Reversed[len(first) - i]

print(maxFinal)


